File: javascript/src/com/example/script/primary.js
// 引入 ToolB 类，相对于src目录的路径
let ToolB = require('com/example/tool/ToolB');
// 创建 ToolB 实例
const toolBInstance = new ToolB();
// 调用 methodB 方法
toolBInstance.methodB($console);
$composeView.create((slot) => {
    // 1. 创建一个状态
    let state = slot.mutableStateOf(0)
    // 创建一个垂直布局
    slot.Column((column) => {
        // 设置垂直布局修饰器
        column.setModifier((m) => {
            // 设置垂直布局填充满屏幕
            m.fillMaxSize(1.0)
        })
        // 水平居中对齐元素
        column.setHorizontalAlignment((alignments) => { return alignments.getCenterHorizontally() })
        // 垂直居中排列元素
        column.setVerticalArrangement((arrangements) => { return arrangements.getCenter() })
        column.setContent((columnSlot) => {
            columnSlot.Text((text) => {
                // 2. 追踪状态变化, 当状态变化时，文本会重新组合
                text.trackSingleState(state)
                // 3. 使用状态值设置文本内容
                text.setText("点击次数: " + state.getValue())
            })
            columnSlot.TextButton((button) => {
                // 设置按钮显示的内容
                button.setContent((buttonSlot) => {
                    // 设置按钮文本
                    buttonSlot.Text((text) => {
                        text.setText("点击我")
                    })
                })
                button.setOnClick(() => {
                    // 4. 点击按钮时，更新状态值
                    state.setValue(state.getValue() + 1)
                })
            })
        })
    })
})
// 启动 android 活动用于展示脚本UI
$activity.start()

File: python/src/python/com/example/script/primary.py
# 必须使用 `from m8test_java.com.m8test.script.GlobalVariables import xxx` xxx是全局变量名
from m8test_java.com.m8test.script.GlobalVariables import _activity
from m8test_java.com.m8test.script.GlobalVariables import _composeView
from m8test_java.com.m8test.script.GlobalVariables import _console

# from 引入时必须使用相对于src目录的路径
from python.com.example.tool.ToolB import ToolB

toolB = ToolB()
toolB.methodB(_console)

# 使用 Compose UI 创建界面
_composeView.create(
    # 创建一个垂直方向的布局
    lambda slot: slot.Column(
        lambda column: (
            # 设置 Column 修饰符, 使其填满整个屏幕
            column.setModifier(lambda modifier: modifier.fillMaxSize(1.0)),
            # 设置 Column 内部子元素的对齐方式为水平居中对齐
            column.setHorizontalAlignment(
                lambda alignments: alignments.getCenterHorizontally()
            ),
            # 设置 Column 内部子元素在垂直方向的排列方式为居中排列
            column.setVerticalArrangement(
                lambda arrangements: arrangements.getCenter()
            ),
            # 设置 Column 的内容
            column.setContent(
                lambda columnSlot: (
                    # 1. 使用 mutableStateOf 创建可变状态
                    (state := columnSlot.mutableStateOf(0)),
                    # 在 Text 组件中显示点击次数
                    columnSlot.Text(
                        lambda text: (
                            # 2. 使用 trackSingleState 追踪状态变化, 如果状态值改变的话, 该 Composable 会重新组合
                            text.trackSingleState(state),
                            # 3. 使用 state.getValue() 获取状态值
                            text.setText("点击次数" + str(state.getValue())),
                        )
                    ),
                    # 创建一个按钮，点击按钮时修改状态值
                    columnSlot.TextButton(
                        lambda textButton: (
                            # 设置按钮内容显示一个简单的文本
                            textButton.setContent(
                                lambda buttonSlot: buttonSlot.Text(
                                    lambda text: text.setText("点击")
                                )
                            ),
                            textButton.setOnClick(
                                # 4. 使用 state.setValue(newValue) 修改状态值
                                lambda: state.setValue(state.getValue() + 1)
                            ),
                        )
                    ),
                )
            ),
        )
    )
)
# 启动ui界面
_activity.start()
